Яценко Ирина

lholypeachy opened the pull request:
@dmnovikova

lholypeachy committed:
Реализация основной задачи

lholypeachy committed:
Исправление теста на плотное размещение в окружности

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,50 @@
+{
+    public class CircularCloudLayouter : ICircularCloudLayouter
+    {
+        public Point Center { get; set; }
```
давай Center и Rectangles вынесем в класс Cloud

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,50 @@
+    {
+        public Point Center { get; set; }
+        public List<Rectangle> Rectangles = new List<Rectangle>();
+        private double _angle;
```
зачем тут наименование с подчеркиванием? обычно '\_' добавляют в начало зарезервированных имен, например, _event_. чтобы компилятор понял, что это название переменной, а не служебное слово. Или иcпользуют в одиночку ( _ ) когда, переменная не нужна
например, метод GetAppleAndBanana() возвращает сразу два значения, одно из которых в конкретном месте вызова не нужно (а в других нужны оба, иначе зачем возвращать два:) )
тогда будет так: `var (apple, _) = GetAppleAndBanana();`

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,50 @@
+            while (HaveIntersection(rectangle)) rectangle.Location = FindNewRectanglePosition();
+
+            Rectangles.Add(rectangle);
+            return rectangle;
```
общепринято оставлять пустую строку перед `return`, это повышает читаемость кода, т.к. сразу можно "выцепить" взглядом, что возвращается

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,50 @@
+
+            var rectangle = new Rectangle(FindNewRectanglePosition(), rectangleSize);
+
+            while (HaveIntersection(rectangle)) rectangle.Location = FindNewRectanglePosition();
```
вкусовшина, но на мой взгляд тело while лучше унести в другую строку, опять же для лучшего чтения кода + лучше сразу обособлять фигурными скобками, т.к. если тело будет разрастаться, без скобок повышается шанс что-то потерять и посадить баг

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,50 @@
+            return rectangle;
+        }
+
+        public Point FindNewRectanglePosition(double deltaAngle = 0.1)
```
сейчас код не очень расширяем, что если разработчик захочет уметь выбирать другой алгоритм раскладки?

давай создадим класс, который будет итеративно возвращать набор точек, потенциальных для расположения прямоугольника (в данном случае на спирали), и передавать в layouter интерфейс

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,50 @@
+        }
+
+        public Point FindNewRectanglePosition(double deltaAngle = 0.1)
+        {
```
почему сейчас модификатор доступа у метода public, если он используется только внутри класса?

dmnovikova replied:
увидела, что используется еще в тестах
открывать метод только для использования в тестах — плохая практика
в разработки следует придерживаться подхода: у тебя есть программа, и все решения ты принимаешь только исходя из ее конструкции, тесты ты пишешь на то, что есть, ничего не меняя в основной программе

dmnovikova replied:
а еще стоит обращать внимание на подсказки решарпера (если вдруг нет — обязательно установи)
обычно то, что он "подсвечивает" нужно менять, или явно понимать, почему нужно оставить так

dmnovikova commented on cs/TagsCloudVizuaizationTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.CodeDom;
```
лишний юзинг

dmnovikova commented on cs/TagsCloudVizuaizationTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,100 @@
+using TagsCloudVisualization;
+using FluentAssertions;
+
+namespace TagsCloudVizualizationTest
```
неверный неймспейс

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,50 @@
+
+        public Rectangle PutNextRectangle(Size rectangleSize)
+        {
+            if (rectangleSize.Width <= 0
```
придирка: тут не такая длинная строка, чтобы на две делить

dmnovikova commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,50 @@
+                Center.X + (int)(Math.Cos(_angle) * radius),
+                Center.Y + (int)(Math.Sin(_angle) * radius));
+
+            return position;
```
сейчас первое слово размещается не по центру облака — в центре находится верхний левый угол прямоугольника, давай будем размещать центр

dmnovikova commented on cs/TagsCloudVizuaizationTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,93 @@
+        public void WhenPutNewRectangle_ShouldBeAddedToList()
+        {
+            layouter.PutNextRectangle(new Size(4, 2));
+
```
их не видно, но они есть :)
тут лишние пробелы, не забывай форматировать файлы перед отправкой на ревью (_ctrl+alt+l_ или _ctrl+alt+shift+l_ в решарпере)

lholypeachy committed:
Исправление стилистических ошибок и опечаток

lholypeachy committed:
Изменила название проекта с тестами

lholypeachy committed:
Создала класс Cloud и отредактировала тесты

Поменяла модификаторы доступа для методов которые используются только внутри класса CircularCloudLayouter

lholypeachy committed:
Добавила новый интерфейс и класс Spiral

lholypeachy committed:
Реализация 2 и 3 задачи

lholypeachy committed:
Добавление изображений в Readme.md

lholypeachy committed:
Добавила тесты для класса Spiral

dmnovikova commented on cs/TagsCloudVisualization/Cloud.cs:
```diff
@@ -0,0 +1,16 @@
+        }
+
+        public readonly Point Center;
+        public List<Rectangle> Rectangles = new();
```
readonly

dmnovikova commented on cs/TagsCloudVisualizationTest/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,99 @@
+
+        [TestOf(nameof(CircularCloudLayouter.PutNextRectangle))]
+        [TestCaseSource(nameof(RectanglesPosition))]
+        public bool WhenPassSeveralRectangles_ShouldReturnCorrectIntersectionResult(Size rectangleSize,
```
если какие-то параметры переносятся на новую строку, то нужно переносить все + каждый должен быть с новой строки : 
```
public bool WhenPassSeveralRectangles_ShouldReturnCorrectIntersectionResult(
    Size rectangleSize,
    Size newRectangleSize)
```
а иногда можно пренебречь длиной

dmnovikova commented:
Реализация 2 и 3 задачи -- плохое название коммита
Представь, что другие люди будут развивать твой проект. Когда им потребуется разобраться в непонятном куске кода, то вероятно, они попробуют обратиться к истории коммитов. Откуда им знать, что ты подразумевала под 2 и 3 задачами полгода/год/пять лет назад?

Лучше в название закладывать фичу, добавленную в коммите: "добавила визуализацию облака ..."

dmnovikova commented on cs/TagsCloudVisualization/TagCloudCreator.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace TagsCloudVisualization
+{
+    public class TagCloudCreator
```
плохое название, т.к. класс не создает облако, а отрисовывает его

dmnovikova replied:
на мой взгляд, visualizer или drawer более очевидные названия

dmnovikova commented on cs/TagsCloudVisualization/TagCloudCreator.cs:
```diff
@@ -0,0 +1,20 @@
+    {
+        public void CreateCloud(RectangleF[] rectangles, string imageName)
+        {
+            var bitmap = new Bitmap(1000, 1000);
```
давай размеры изображение передавать через параметры или через конструктор (старайся нигде кроме Program не хардкодить никакие значения)

не обязательно делать: можно докрутить решение и передавать visualizingSettings, в которых будут содержаться размеры, возможные цвета для отрисовки и тд

dmnovikova commented on cs/TagsCloudVisualization/TagCloudCreator.cs:
```diff
@@ -0,0 +1,20 @@
+            var pen = new Pen(Color.Crimson);
+
+            graphics.Clear(Color.Black);
+            graphics.DrawRectangles(pen, rectangles);
```
сейчас тут проблема: как отрисуется облако, если будет больше, чем размеры изображения?
нужно добавить масштабирование размеров облака относительно размеров изображения

lholypeachy committed:
Добавила масштабирование размеров облака относительно размеров изображения

dmnovikova commented:
Зачтено ✅
